### PR TITLE
Fix serialization data contract overflow when reading object concurrently

### LINKED ISSUE #4504: Data Contract overflow when reading objects concurrently
Data contract serializers are prone to cache exception when used by multiple threads. This is an issue that are being fixed on Desktop which should be ported to CoreClr.

### PR DESCRIPTION
**Pull request by khdang**
This change is to port the fix from Desktop (CS 1558305) which has been validated.

@SGuyGe @shmao

Fix #4504

### CODE REVIEW

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContract.cs [hunk 1]
```diff
@@ -657,55 +658,67 @@ internal static int GetId(RuntimeTypeHandle typeHandle)
             // check whether a corresponding update is required in ClassDataContract.IsNonAttributedTypeValidForSerialization
             private static DataContract CreateDataContract(int id, RuntimeTypeHandle typeHandle, Type type)
             {
-                lock (s_createDataContractLock)
+                DataContract dataContract = s_dataContractCache[id];
```
Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContract.cs]: It looks like s_dataContractCache might be read concurrently with being written, as the lock isn't taken on all accesses to s_datacontractCache, only when writing to it and when the initial check fails.  That is not a safe thing to do with a ```Dictionary<...>```, which is not thread-safe for a read concurrently with a write.

The comment on this PR indicates this is a port of a change from the full framework.  Is the full framework using a ```Hashtable``` or a ```Dictionary<...>```?  Hashtable is actually safe for reading concurrently with a single write.

Comment (khdang) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContract.cs]: The full framework is using ```Dictionary<...>``` for this fix, which was made by @jcondex. I think this is a targeted fix for the scenario of concurrent deserialization. There was an iteration of this fix which uses locks more properly but that introduces performance regression in common scenarios.

Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContract.cs]: > There was an iteration of this fix which uses locks more properly but that introduces performance regression in common scenarios.

That's why ```ConcurrentDictionary<...>``` exists.

> The full framework is using Dictionary<...> for this fix

In that case it appears to be flawed in this regard as well, unless there's something else I'm missing that's making it safe.

Comment (khdang) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContract.cs]: My bad I looked at the wrong one. The s_dataContractCache is an array and not a ```Dictionary<...>```. The id parameter was computed in DataContractCriticalHelper.GetId which utilizes a lock so I think this is thread-safe.

**Comment by stephentoub** [comment]
Ah, ok, I was also looking at wrong one.

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by shmao** [comment]
LGTM.

**Comment by SGuyGe** [comment]
LGTM

**Comment by stephentoub** [comment]
Test Innerloop CentOS7.1 Debug Build and Test please

**Pull request merged**